File: SDE_sheet/Arrays/22_largest_subarray_with_kSum/solution.py
class Solution:
    def maxLen(self, n, arr):
        sum = 0
        max_len = 0
        map = {}
        
        for i, num in enumerate(arr):
            sum += num
            if sum == 0:
                max_len = i + 1
            if sum in map:
                max_len = max(max_len, i - map[sum])
            if sum not in map:
                map[sum] = i
        
        return max_len

File: SDE_sheet/Arrays/22_largest_subarray_with_kSum/solution.cpp
class Solution {
  public:
    int maxLen(vector<int>& arr, int n) {
       int sum = 0, max = 0;
        std::unordered_map<int, int> map;
        
        for (int i = 0; i < arr.size(); ++i) {
            sum += arr[i];
            if (sum == 0) {
                max = i + 1;
            }
            if (map.find(sum) != map.end()) {
                max = std::max(max, i - map[sum]);
            }
            if (map.find(sum) == map.end()) {
                map[sum] = i;
            }
        }
        return max;
    }
};
